docs/lang/articles/advanced/sparse.md

```diff
@@ -2,100 +2,116 @@
 sidebar_position: 3
 ---
 
-# Sparse computation
-
-Compiler-level support for spatially sparse computation is a unique feature of Taichi.
-
-![image](https://raw.githubusercontent.com/taichi-dev/public_files/master/taichi/sparse_grids.gif)
-
-Figure: A swinging "Taichi" pattern represented with a 512x512 sparse grid. The sparse grid has a multi-level *tree* structure.
-White stands for inactive tree nodes, and active tree nodes are darker.
-
-The sparse grid above has the following structure:
-- The grid is divided into 8x8 `block1` containers;
-- Each `block1` container has 4x4 `block2` cells;
-- Each `block2` container has 4x4 `block3` cells;
-- Each `block3` container has 4x4 pixel cells;
-- Each pixel contains an `i32` value `x[i, j]`.
+# Sparse spatial data structures
 
 :::note
-For more information about *cells* and *containers*, see [**Data structure organization**](../misc/internal.md#data-structure-organization).
-In this article, you can assume *containers* and *cells* are the same.
+Pre-request: please read the [SNodes](lang/articles/misc/internal.md#data-structure-organization) first.
 :::
 
-Taichi allows you to define sparse data structures effortlessly. For example, you can define the grid above as
-
-```python
-x = ti.field(dtype=ti.i32)
-
-block1 = ti.root.pointer(ti.ij, 8)
-block2 = block1.pointer(ti.ij, 4)
-block3 = block2.pointer(ti.ij, 4)
-block3.dense(ti.ij, 4).place(x)
-```
-[[Full source code of this animation]](https://github.com/taichi-dev/taichi/blob/master/python/taichi/examples/features/sparse/taichi_sparse.py)
-
-Intuitively, a sparse grid in Taichi allows you to use memory space more wisely, since only tree nodes involved in computation are allocated.
-Now, let's take a step back and think about *why we need sparse grids, how to define them in Taichi, and how to compute on these data structures*.
-
 ## Motivation
 
 High-resolution 2D/3D grids are often needed in large-scale spatial computation, such as physical simulation, rendering, and 3D reconstruction.
-However, these grids tend to consume a huge amount of memory space and computation.
+However, these grids tend to consume a huge amount of memory space and computation if we use dense data structure (see [field](lang/articles/basic/field.md) and [field advanced](lang/articles/advanced/layout.md)).
 While a programmer may allocate large dense grids to store spatial data (especially physical quantities such as a density or velocity field),
 oftentimes, they only care about a small fraction of this dense grid since the rest may be empty space (vacuum or air).
 
-In short, the regions of interest in sparse grids may only occupy a small fraction of the whole bounding box.
+<center>
+
+![BVH](https://github.com/FantasyVR/public_files/blob/sparse_computation/taichi/doc/bvh_small.png)
+
+</center>
+
+As shown above, the regions of interest in sparse grids may only occupy a small fraction of the whole bounding box.
```

Comment:
Shall we add some transition phrases like "for example" to clearly separate it from the context?